docs/adr/0020-application-architecture.md

```diff
@@ -0,0 +1,110 @@
+- Adding new objects needed in a route handler can be time-consuming as the object must be plumbed through the entire initialization chain vs. more elegant methods like Dependency Injection (DI).
+- Not clear where/how to add new components and takes time to study/understand how things are currently setup in an attempt to mimic the structure for the new component.
+- Time consuming to setup boiler-plate for components, as we have no tooling to work with the current ad-hoc application architectures.
+- Our ad-hoc architecture frequently mixes concerns such as having business logic mixed in with request handling logic, and has other warts from its evolution over time vs. being planned up front.
+- New back-end services evolve differently resulting in more ad-hoc application architectures to learn.
+- Shared components in `fxa-shared` can't rely on basic object lifecycles or setup approaches as they may be used in multiple different ad-hoc application architectures.
+
+Not choosing an application framework means that we have choosen to make ad-hoc application architectures which will continue to exhibit the problems above.
+
+It is assumed that the four newest FxA back-end services (admin-server, support-panel, event-broker, gql-api) will be switched to the chosen approach for consistency.
+
+## Decision Drivers
+
+- Documented application architecture.
+- Tooling that reduces boiler-plate and creates consistent code architecture.
+- Modern paradigms to ease creation of global objects and their testability, such as DI.
+- Training materials so that new developers can understand and work on application features easily.
+- Handles situations we need such as:
+  - Exception handling
+  - Validation
+  - Service instantiation (DI, etc)
+  - Authentication
+  - RESTful API's (Swagger/OpenAPI a bonus)
+  - GraphQL
+  - Use of components/services easily in a script
+
+## Considered Options
+
+- Evolve/improve our ad-hoc application architecture with docs, tools, etc.
+- Use an existing well-documented and popular application framework such as:
+  - NestJS
+  - Adonis
+  - LoopBack
+
+## Decision Outcome
+
+### Positive Consequences
+
+### Negative Consequences
+
+## Pros and Cons of the Options
+
+### Evolve/improve our ad-hoc application architecture with docs, tools, etc.
+
+Our existing ad-hoc structure typically involves a few chosen libraries and a basic directory structure. Error handling differs in several of the back-end services, while validation is always provided by the `joi` library. To document our ad-hoc application architecture we would first need to make clear choices about what libraries we use and how their use should be structured in the application. We could then write tooling and documentation around these choices.
+
+- Pros:
+  - Could be easier to convert our existing backend services as they will likely be closer in libs/structure to the choosen structure if desired.
```

Comment:
Agreed, that is definitely a notable pro. Though I think its feasible to back-port patterns/conventions from something like NestJS into our existing apps (like auth-server) so they can take advantage of the conventions even if they're not using the library.

An example:
In Nest, there's the convention that data transfer objects (DTO's) are in files, named as such along with the code they go to. While the routes are in a controller file. Business logic typically goes in a service. These filenames end like so:
- `*.dto.ts`
- `*.controller.ts`
- `*.service.ts`

So I could imagine copying the convention into auth-server for Stripe/subscription routing as being something like:
- Move URL handlers from `routes/subscriptions.ts` to `payments/payments.controller.ts`
- Consolidate business logic from `routes/subscriptions.ts` and `payments/stripe.ts` to `payments/stripe.service.ts`
- Move DTO's from `routes/validators.js` to `routes/dto/*.dto.ts`
- Move tests into `*.spec.ts` files that sit alongside what they test (as we use in front-end TS projects, and as Nest also uses)

Doing something like this we'd gain the advantage that even though some of the tooling doesn't work for our legacy projects, we can at least align them on other conventions that would make their style more inline with a new NestJS project.

Unfortunately we can't use any of these 3 full frameworks DI scheme without using the rest of the framework, so we might just have to use a standalone DI like `typedi` to benefit from that. 

This might be a good hybrid approach that lets us align the existing code-bases with NestJS naming/organization conventions while going completely to NestJS for the new services.